sample/buildSrc/src/main/kotlin/com/dropbox/sample/tasks/AffectedTasksPlugin.kt

```diff
@@ -0,0 +1,130 @@
+                project.extensions.findByName(TEST_TASK_TO_RUN_EXTENSION)
+            ) as TestTasks
+            registerAffectedTestTask(
+                "runAffectedUnitTests",
+                testTasks.jvmTest, testTasks.jvmTestBackup, rootProject,
+            )
+
+            registerAffectedAndroidTests(rootProject)
+            registerAffectedConnectedTestTask(rootProject)
+        }
+
+        filterAndroidTests(project)
+        filterUnitTests(project)
+    }
+
+    private fun registerAffectedTestTask(
+        taskName: String, testTask: String, testTaskBackup: String?,
+        rootProject: Project
+    ): Task {
+        val task = rootProject.tasks.register(taskName) { task ->
+            val paths = getAffectedPaths(testTask, testTaskBackup, rootProject)
+            paths.forEach { path ->
+                task.dependsOn(path)
+            }
+            task.enabled = paths.isNotEmpty()
+            task.onlyIf { paths.isNotEmpty() }
+        }
+        return task.get()
+    }
+
+    private fun getAffectedPaths(
+        task: String,
+        taskBackup: String?,
+        rootProject: Project
+    ):
+            Set<String> {
+        val paths = LinkedHashSet<String>()
+        rootProject.subprojects { subproject ->
+            val pathName = "${subproject.path}:$task"
+            val backupPath = "${subproject.path}:$taskBackup"
+            if (AffectedModuleDetector.isProjectProvided(subproject)) {
+                if (subproject.tasks.findByPath(pathName) != null) {
+                    paths.add(pathName)
+                } else if (taskBackup != null &&
+                    subproject.tasks.findByPath(backupPath) != null
+                ) {
+                    paths.add(backupPath)
+                }
+            }
+        }
+        return paths
+    }
+
+    private fun registerAffectedConnectedTestTask(
+        rootProject: Project
+    ) {
+        registerAffectedTestTask(
+            "runAffectedAndroidUnitTests",
+            testTasks.runAndroidTestTask, null, rootProject
+        )
+    }
+
+    private fun registerAffectedAndroidTests(
+        rootProject: Project
+    ) {
+        registerAffectedTestTask(
+            "assembleAffectedAndroidTests",
+            testTasks.assembleAndroidTestTask, null, rootProject
+        )
+    }
+
+
+    private fun filterAndroidTests(project: Project) {
+        val tracker = DependencyTracker(project, null)
+        project.tasks.all { task ->
+            if (task.name.contains(ANDROID_TEST_BUILD_VARIANT)) {
+                tracker.findAllDependents(project).forEach { dependentProject ->
+                    dependentProject.tasks.forEach { dependentTask ->
+                        AffectedModuleDetector.configureTaskGuard(dependentTask)
+                    }
+                }
+                AffectedModuleDetector.configureTaskGuard(task)
+            }
+        }
+    }
+
+    private fun filterUnitTests(project: Project) {
```

Comment:
doesn't this filter capture all the `AndroidTest` tasks as well?

Reply:
Unfortunately, no - `assemble*AndroidTest` extends from a `DefaultTask` and `connected*AndroidTest` extends from another class (which I can't recall at the moment).